devito/ir/cgen/printer.py

```diff
@@ -33,25 +33,101 @@ class CodePrinter(C99CodePrinter):
+                         **CodePrinter._default_settings}
+
+    _func_prefix = {}
+    _func_litterals = {}
+
+    _qualifiers_mapper = {
+        'is_const': 'const',
+        'is_volatile': 'volatile',
+        '_mem_constant': 'static',
+        '_mem_shared': '',
+    }
+
+    _prec_litterals = {np.float32: 'F', np.complex64: 'F'}
+
+    _restrict_keyword = 'restrict'
+
+    _includes = []
+
+    _namespaces = []
+
+    _headers = [('_POSIX_C_SOURCE', '200809L')]
 
     @property
     def dtype(self):
-        return self._settings['dtype']
+        try:
+            return self._settings['dtype'].nptype
+        except AttributeError:
+            return self._settings['dtype']
 
     @property
     def compiler(self):
-        return self._settings['compiler']
+        return self._settings['compiler'] or configuration['compiler']
+
+    def doprint(self, expr, assign_to=None):
+        """
+        The sympy code printer does a lot of extra we do not need as we handle all of
+        it in the compiler so we directly defaults to `_print`.
+        """
+        return self._print(expr)
 
-    def single_prec(self, expr=None):
-        dtype = sympy_dtype(expr) if expr is not None else self.dtype
-        return dtype in [np.float32, np.float16]
+    def _prec(self, expr):
+        try:
+            dtype = sympy_dtype(expr, default=self.dtype)
+        except TypeError:
+            return self.dtype
+        if dtype is None or np.issubdtype(dtype, np.integer):
+            if any(isinstance(i, Float) for i in expr.atoms()):
+                try:
+                    return np.promote_types(self.dtype, np.float32).type
+                except np.exceptions.DTypePromotionError:
+                    # Corner cases, e.g. Void, cannot (shouldn't) be promoted
+                    return self.dtype
+            else:
+                return dtype or self.dtype
+        else:
+            return dtype or self.dtype
+
+    def prec_literal(self, expr):
+        return self._prec_litterals.get(self._prec(expr), '')
+
+    def func_literal(self, expr):
+        return self._func_litterals.get(self._prec(expr), '')
+
+    def func_prefix(self, expr, mfunc=False):
+        prefix = self._func_prefix.get(self._prec(expr), '')
+        if mfunc:
+            return prefix
+        else:
+            return '' if prefix == 'f' else prefix
 
     def parenthesize(self, item, level, strict=False):
         if isinstance(item, BooleanFunction):
-            return "(%s)" % self._print(item)
+            return f"({self._print(item)})"
         return super().parenthesize(item, level, strict=strict)
 
+    def _print_PyCPointerType(self, expr):
+        ctype = f'{self._print_type(expr._type_)}'
+        if ctype.endswith('*'):
+            return f'{ctype}*'
+        else:
+            return f'{ctype} *'
+
+    def _print_type(self, expr):
+        try:
+            expr = dtype_to_ctype(expr)
+        except TypeError:
+            pass
+        try:
```

Comment:
I suspect there's an issue here. If you end up in the `except` above, `expr` won't have been defined at this point. But then you use it as key for `type_mappings`... so either the try-except above is useless, or I don't know

Reply:
`expr` is the input arg so it's defined yes

Reply:
oh... right ... I was 😴 . Fine